### PR TITLE
Move to keyset pagination

### LINKED ISSUE #214: Reduce sorting memory resource requirements
When using `join=True` and a data source which is larger than a gigabyte `sort_output=True` appears to cause a large memory spike during join operations. This may cause a system crash in scenarios where not enough memory is available.

### PR DESCRIPTION
**Pull request by d33bs**
# Description

This PR modifies "chunk" behavior to follow keyset pagination patterns. We gather sets of pages as `pagesets` which are based on a key per table which is used to help create pages of rows from tables. `pagesets` are gathered ahead of time to retain the ability to parallelize where appropriate. This change required a new configuration parameter to be specified per table and the overall join.

Operationally this reduces the amount of data stored in tables throughout, reduces the complexity of the sorting performed, and avoids unnecessary data loading which occurs during the use of `OFFSET`.

Changes here made it unnecessary to create metadata columns and as a result I've removed those from all spots I could find.

Changes here are intended to address #214 , which turned out to be more related to data size than the sorting implementation. This is confirmed by a now passing `large_data_tests` test of JUMP data using the full dataset (instead of a truncated version).

Closes #214

## What is the nature of your change?

- [x] Bug fix (fixes an issue).
- [ ] Enhancement (adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).
- [ ] This change requires a documentation update.

# Checklist

Please ensure that all boxes are checked before indicating that a pull request is ready for review.

- [x] I have read the [CONTRIBUTING.md](CONTRIBUTING.md) guidelines.
- [x] My code follows the style guidelines of this project.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] New and existing unit tests pass locally with my changes.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have deleted all non-relevant text in this pull request template.

### CODE REVIEW

File: cytotable/convert.py
Comment (falquaddoomi) [inline cytotable/convert.py:299]: I don't see `page_keys` in the list of arguments to this function or any references to it in the body.

File: cytotable/convert.py
Comment (falquaddoomi) [inline cytotable/convert.py:285]: I see that the type here is `Tuple[int, int]`; does that imply that columns identified in `page_keys` are also integers? If so, you might want to add that to the documentation and/or add a check that they are in fact integers.

File: cytotable/convert.py
Comment (falquaddoomi) [inline cytotable/convert.py:178]: It seems that `page_key` needs to be orderable and from the definition of `pagesets` elsewhere I presume it also has to be an integer, but is it the case that it also has to be unique? If there are repeats, I imagine you'd end up with page sizes larger than the chunk size you specified. Perfectly fine if the page key needs to be unique, but that's something I'd add to the documentation if so.

You might also consider tightening the return type from `List[Any]` to `List[Tuple[int,int]]` to be more explicit about the chunks being a list of integer key pairs that define each page.

File: cytotable/convert.py
Comment (falquaddoomi) [inline cytotable/convert.py:276]: I think this code is fine as-is, but for brevity you might consider using the `grouper()` function as described in the [itertools docs](https://docs.python.org/3.10/library/itertools.html) -- unfortunately I can't link to the function directly, but you'll find it if you search for "def grouper".

A simplified version of that function, [`chunked()`](https://more-itertools.readthedocs.io/en/stable/api.html#more_itertools.chunked), is included in the `more-itertools` library, but I don't know if you want to take an extra dependency for such a simple function.

You could use the `chunked()` function like so to produce your set of non-overlapping intervals:
```python
chunks = [(x[0], x[-1]) for x in chunked(keys, chunk_size)]
```

If you instead use `grouper()`, you'll have to account for the case in which a group ends up padded with `None` if it can't construct a group of size `chunk_size`.
```python
[(x[0], x[-1] if x[-1] is not None else x[0]) for x in grouper(keys, chunk_size)]
```

It seems that SQL's `BETWEEN` is inclusive, so a page like `(100, 100)` will still match the keys with value `100`.

**Review by falquaddoomi — COMMENTED** [review]
Neat solution to the problem you were facing; I agree that it's less brittle than using `LIMIT` and `OFFSET` since you're now relying on the data rather than implementation-specific definitions of `LIMIT` and `OFFSET`.

I think you could add a bit more description of what columns are good candidates for keys; from my reading of your code I assume they'd be unique integers.

Anyway, nice work, and glad to see this issue put to rest!

File: cytotable/convert.py
Comment (d33bs) [reply cytotable/convert.py:299]: Thanks great catch, I'll update accordingly!

File: cytotable/convert.py
Comment (d33bs) [reply cytotable/convert.py:285]: Thanks, good point - as designed `page_keys` need to be numeric (ints or floats). I'll update the type here along with the documentation aspects.

File: cytotable/convert.py
Comment (d33bs) [reply cytotable/convert.py:178]: Thanks good point! Yes, there are some assumptions we should document upfront for the `pagesets`. I tried to be loose in the considerations here but there are a few gotchas which resulted in a more complicated implementation of the `pageset` list of tuples implementation (will touch on this in the other comment).

File: cytotable/convert.py
Comment (d33bs) [reply cytotable/convert.py:276]: Thanks I'll explore this a bit more! I found that `grouper()` was challenging to use with concerns to overlaps that should respect the chunk size. There's likely a more elegant solution as you suggested.

Comment (d33bs) [reply cytotable/convert.py:276]: Thanks again. I reworked this to simplify and demonstrate reasoning. I decided to create a function out of this work and test it individually to ensure we know what we're receiving (and why the implementation doesn't use existing algorithms in built-in's). My hope here is that the test cases help demonstrate better what output we're expecting (it's a bit different than `grouper`, `groupby`, or `more_itertools.chunked` provide based on repeated value and SQL-specific expectations).

**Comment by d33bs** [comment]
Thanks @falquaddoomi for the review and comments! I made some adjustments based on what you mentioned.

File: cytotable/convert.py
Comment (jenna-tomkinson) [inline cytotable/convert.py:210]: Why do imports happen at two points in this file? I could be mistaken, but isn't it more standard to have all imports at the start of the file?

File: cytotable/convert.py
Comment (jenna-tomkinson) [inline cytotable/convert.py:47]: Why is pathlib removed?

File: cytotable/convert.py
Comment (jenna-tomkinson) [inline cytotable/convert.py:266]: Should this function be renamed to not use chunk since based on what I can see, you don't use chunk_size anymore? Would `_source_pageset_to_parquet` be more accurate?

File: cytotable/convert.py
Comment (jenna-tomkinson) [inline cytotable/convert.py:322]: How are metadata columns added now with this change?

File: tests/test_convert_threaded.py
Comment (jenna-tomkinson) [inline tests/test_convert_threaded.py:129]: Was this deleted because of the pagination fix?

File: cytotable/convert.py
Comment (falquaddoomi) [reply cytotable/convert.py:210]: That's generally a good point, but I believe this is because the parallelization library that @d33bs is using, [parsl](https://parsl-project.org/), requires that functions that are parallelized import all the libraries they need in the function, not within the enclosing module like you'd normally do. You can tell it's an entrypoint from the `@python_app` decorator before the function definition. 

Here's a bit from the parsl docs at least saying it's necessary to import within parallelized functions: https://parsl.readthedocs.io/en/stable/1-parsl-introduction.html#Python-Apps.

File: cytotable/constants.py
Comment (d33bs) [reply cytotable/constants.py:76]: Thanks great question! These columns were added to help implement offset-based pagination (avoiding the use of existing table columns). I opted to remove them as keyset pagination focuses on using existing table keys for pagination. The removal also reduces the complexity of the code more generally which likely has performance improvements (we're adding 3 fewer columns to each table and each chunk of each table).

File: cytotable/convert.py
Comment (d33bs) [reply cytotable/convert.py:210]: Thanks Jenna for a good question and Faisal for the docs. Concurring with what Faisal mentioned, this is to help with multiprocessing through Parsl where each decorated function might not have access of the full module's imports.

File: cytotable/convert.py
Comment (d33bs) [reply cytotable/convert.py:47]: Thanks - `pathlib` was removed as I noticed it was no longer needed for `_get_table_columns_and_types`. I believe the import inadvertently remained after earlier changes than this PR, so this was a bit of opportunistic refactoring (I noticed it during the work towards this PR).

File: cytotable/convert.py
Comment (d33bs) [reply cytotable/convert.py:266]: Thanks, great point - I like that change and will make a further revision to avoid the use of "chunk" to defer to "page" where possible.

File: cytotable/convert.py
Comment (d33bs) [reply cytotable/convert.py:322]: Thanks for asking about this, CytoTable-specific metadata columns (columns which add CytoTable data) are no longer added to tables. Metadata columns from the source data (for ex. from CellProfiler) will remain unchanged in terms of how they are processed by CytoTable.

File: tests/test_convert_threaded.py
Comment (d33bs) [reply tests/test_convert_threaded.py:129]: Thanks - yes, this is removed to defer to the preset default SQL for joins. The change here is mostly to remove `LIMIT 5000` from each joined table (from Parquet datasets) and instead use the full dataset. With this change we are now able to process the full dataset within the scope of a [GitHub Actions runner image's resources](https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners/about-github-hosted-runners#standard-github-hosted-runners-for-public-repositories) (4 CPU, 16 GB RAM). Earlier changes towards offset-based pagination resulted in #214 and an inability to run the full data through this test. Keyset pagination implemented in this PR is much more performant and doesn't require us to artificially limit the query here anymore.

File: cytotable/convert.py
Comment (d33bs) [reply cytotable/convert.py:266]: Thinking about this more, I'm going to make minimal changes to the usage of those words. Generally I feel we need to distinguish between Parquet pages vs CytoTable pages, and retaining "chunk" might be more understandable for now (in addition to avoiding a change to the `convert` API.

File: cytotable/convert.py
Comment (gwaybio) [inline cytotable/convert.py:231]: is there a better name for this variable? maybe `page_keys`?

File: cytotable/convert.py
Comment (gwaybio) [inline cytotable/convert.py:259]: same comment on variable name here

File: cytotable/convert.py
Comment (gwaybio) [inline cytotable/convert.py:352]: I'm likely not reading this correctly, but is this ordering by `page_key` for every chunked/pageset operation? Wouldn't it be more efficient to order after all pagesets are processed?

File: cytotable/convert.py
Comment (gwaybio) [inline cytotable/convert.py:759]: Same question about ordering

File: cytotable/convert.py
Comment (gwaybio) [inline cytotable/convert.py:1060]: Can you add a bit more details to this exception? Any further details about how a user might be able to resolve this exception would be helpful. Why might someone encounter this error?

File: cytotable/convert.py
Comment (gwaybio) [inline cytotable/convert.py:1423]: Can you answer the question "why?" in this error message? Or perhaps provide an example?

File: cytotable/utils.py
Comment (gwaybio) [inline cytotable/utils.py:193]: Might be helpful to define the variable without using the variable name. Perhaps something like "The column name to be used to identify and join different pagination chunks"?

File: cytotable/utils.py
Comment (gwaybio) [inline cytotable/utils.py:195]: Similar comment here - this is the chunk identifier for some indicator of where the chunk is in context with the rest of the data?

File: cytotable/utils.py
Comment (gwaybio) [inline cytotable/utils.py:275]: Is this ordering different than the `convert` ordering?

File: cytotable/utils.py
Comment (gwaybio) [inline cytotable/utils.py:594]: Can you switch to active voice? Would a user provide the keys? Would some function generate them?

File: docs/source/overview.md
Comment (gwaybio) [inline docs/source/overview.md:294]: How are pageset keys defined?

Comment (gwaybio) [reply docs/source/overview.md:294]: Ah, I see the example below. It might be super simple to add thi:

"as defined within the _pageset_ keys (see example usage below)."

File: docs/source/overview.md
Comment (gwaybio) [inline docs/source/overview.md:300]: Can you add some comment on defaults? A user would only need to change this...when?

**Review by gwaybio — APPROVED** [review]
Exciting to see this enhancement! I made several comments, mostly aimed at improving clarity in comments and variable names. After addressing, feel free to merge 👍

File: cytotable/convert.py
Comment (d33bs) [reply cytotable/convert.py:231]: Thanks, good thought - I'll get this updated!

File: cytotable/convert.py
Comment (d33bs) [reply cytotable/convert.py:259]: Thanks - will update this as well!

File: cytotable/convert.py
Comment (d33bs) [reply cytotable/convert.py:352]: Thanks, great thought - this was a vestige of the earlier offser-based pagination and I believe we could update here for more performant operations as you mentioned.

File: cytotable/convert.py
Comment (d33bs) [reply cytotable/convert.py:759]: Thanks - will update as mentioned in the prior comment repsonse.

File: cytotable/convert.py
Comment (d33bs) [reply cytotable/convert.py:1060]: Thanks - I'll add more context here and in the exception. In order for keyset pagination to operate the column / key name must be in the table. This exception raises if the column is not found in the table.

File: cytotable/convert.py
Comment (d33bs) [reply cytotable/convert.py:1423]: Thanks - I'll add additional context as you suggested. Because the join composite table may include variable columns contingent on the join SQL there is a special key which may be specified in CytoTable to help perform keyset pagination on the join table. If there is no 'join' key specified when attempting to join we except here to help prevent further operations.

File: cytotable/utils.py
Comment (d33bs) [reply cytotable/utils.py:193]: Thanks, that's more clear - I'll update this accordingly.

File: cytotable/utils.py
Comment (d33bs) [reply cytotable/utils.py:195]: Thanks, yes, this is a "set" of the overall "whole" of all keysets for a given table. I'll add more context here to the documentation.

File: cytotable/utils.py
Comment (d33bs) [reply cytotable/utils.py:275]: Thanks for checking on this - this isn't different from the existing `convert` ordering but was restated to help with SQLite-specific processing here.

Comment (d33bs) [reply cytotable/utils.py:275]: I'm going to move this around a bit in coordination with other comments about sorting.

File: cytotable/utils.py
Comment (d33bs) [reply cytotable/utils.py:594]: Thanks, yes, I'll switch this to active voice. At the moment keys must be provided by a user via the `page_keys` parameter and eventually reach this function to help create "pagesets" which may be processed independently of one another.

File: docs/source/overview.md
Comment (d33bs) [reply docs/source/overview.md:294]: Thanks for this - I'll add further content as you mentioned to help clarify things.

File: docs/source/overview.md
Comment (d33bs) [reply docs/source/overview.md:300]: Thanks, I'll add further content here about when / why this might be changed.

File: cytotable/utils.py
Comment (d33bs) [reply cytotable/utils.py:594]: Looking into this more I realized it might not be possible to reach this exception case given the design of the function and how `convert` works. I ended up removing it instead to keep things simple.

File: cytotable/convert.py
Comment (d33bs) [reply cytotable/convert.py:354]: After looking into this more I felt that it'd be best to leave this line as-is and add additional changes to ensure the sorted results arrive sorted within the concatenation phase of CytoTable. I opted to do this because sorting for individual pages vs entirety of the dataset will have a lower performance impact and be more likely to succeed alongside resource constraints.

The result here will sort each pageset meaning the results within a single range of values are sorted. Afterwards, during concatenation to a single file, the results are combined using a natural sort over the filenames (which include the pageset range) to create a fully sorted dataset per source table and eventually joined table.

File: cytotable/convert.py
Comment (d33bs) [reply cytotable/convert.py:769]: Please reference https://github.com/cytomining/CytoTable/pull/220#discussion_r1731861318 for full description of how this was addressed.

File: cytotable/utils.py
Comment (d33bs) [reply cytotable/utils.py:273]: This was retained in coordination with this answer: https://github.com/cytomining/CytoTable/pull/220#discussion_r1731861318

**Comment by d33bs** [comment]
Thank you @gwaybio, @jenna-tomkinson and @falquaddoomi for your reviews! After making adjustments for all comments I believe this is now ready and I'll merge this in.

**Pull request merged**
